datafusion/optimizer/src/decorrelate_where_exists.rs

```diff
@@ -137,8 +137,14 @@ fn optimize_exists(
     let subqry_inputs = query_info.query.subquery.inputs();
     let subqry_input = only_or_err(subqry_inputs.as_slice())
         .map_err(|e| context!("single expression projection required", e))?;
-    let subqry_filter = Filter::try_from_plan(subqry_input)
-        .map_err(|e| context!("cannot optimize non-correlated subquery", e))?;
+    let subqry_filter = match subqry_input {
```

Comment:
I don't understand why this fixes the error -- what was in the projection in the subquery that caused the problem?

Reply:
I have been looking at this as well. The existing code works for a very simple projection but does not work if the projection is wrapped in any other operator, such as Distinct, Filter, Limit, Sort, and so on.

Reply:
I understand what is happening now and have some suggestions for improving this rule.

This line of code looks at inputs of the subquery and does not care what type of operator the subquery is. Previously this was assumed to be a `Projection` but now it could be a `Projection` or a `Distinct`, or something else ... I think we should add some pattern matching here.

```rust
let subqry_inputs = query_info.query.subquery.inputs();
```

We are then matching on this input and previously expected a `Filter` buit now could be a `Projection` containing a `Filter` because everything is shifted down by one because of the root `Distinct`.

Reply:
I would like to see something with explicit pattern matching to make sure we are only supporting intended cases. Here is my attempt:

```rust
fn optimize_exists(
    query_info: &SubqueryInfo,
    outer_input: &LogicalPlan,
    outer_other_exprs: &[Expr],
) -> datafusion_common::Result<LogicalPlan> {

    let subqry_filter = match query_info.query.subquery.as_ref() {
        LogicalPlan::Distinct(subqry_distinct) => match subqry_distinct.input.as_ref() {
            LogicalPlan::Projection(subqry_proj) => Filter::try_from_plan(&*subqry_proj.input),
            _ => Err(DataFusionError::NotImplemented("todo: error message".to_string()))
        }
        LogicalPlan::Projection(subqry_proj) =>  Filter::try_from_plan(&*subqry_proj.input),
        _ => Err(DataFusionError::NotImplemented("todo: error message".to_string()))
    }.map_err(|e| context!("cannot optimize non-correlated subquery", e))?;
```

Reply:
done, thanks for your advice. @andygrove